--- qemu-2.3.0/cpu-exec.c.orig     2014-12-09 14:45:40.000000000 +0000
+++ qemu-2.3.0/cpu-exec.c  2015-02-20 22:07:02.966000000 +0000
@@ -28,6 +28,12 @@
 #include "exec/memory-internal.h"
 #include "qemu/rcu.h"

+extern abi_ulong baeum_entry_point; /* ELF entry point (_start) */
+extern void baeum_setup(void);
+extern void baeum_log_bb(abi_ulong addr);
+
+#include "../patches/afl-qemu-cpu-inl.h"
+
 /* -icount align implementation. */

 typedef struct SyncClocks {
@@ -296,8 +298,11 @@
     }
  not_found:
    /* if no translated code available, then translate it now */
+
     tb = tb_gen_code(cpu, pc, cs_base, flags, 0);

+    afl_request_tsl(pc, cs_base, flags);
+
  found:
     /* Move the last found TB to the head of the list */
     if (likely(*ptb1)) {
@@ -492,6 +503,13 @@
                     next_tb = 0;
                     tcg_ctx.tb_ctx.tb_invalidated_flag = 0;
                 }
+
+                if(tb->pc == baeum_entry_point) {
+                  afl_forkserver(env);
+                  baeum_setup();
+                }
+                baeum_log_bb(tb->pc);
+
                 if (qemu_loglevel_mask(CPU_LOG_EXEC)) {
                     qemu_log("Trace %p [" TARGET_FMT_lx "] %s\n",
                              tb->tc_ptr, tb->pc, lookup_symbol(tb->pc));
