--- qemu-2.3.0/linux-user/signal.c.orig	2016-07-31 21:59:09.666337987 +0900
+++ qemu-2.3.0/linux-user/signal.c	2016-08-03 20:36:35.469159702 +0900
@@ -30,6 +30,8 @@
 #include "qemu-common.h"
 #include "target_signal.h"
 
+extern void baeum_exit(void);
+
 //#define DEBUG_SIGNAL
 
 static struct target_sigaltstack target_sigaltstack_used = {
@@ -461,6 +463,7 @@
         getrlimit(RLIMIT_CORE, &nodump);
         nodump.rlim_cur=0;
         setrlimit(RLIMIT_CORE, &nodump);
+        baeum_exit();
         (void) fprintf(stderr, "qemu: uncaught target signal %d (%s) - %s\n",
             target_sig, strsignal(host_sig), "core dumped" );
     }
